### DISCUSSION TITLE
What happen to --ignore-optional?

### DISCUSSION THREAD

**Discussion by bodinsamuel** [yarnpkg/berry#2412] [Q&A] [answered]
There used to be a `--ignore-optional` with `yarn install` command.
It was helpful to add optional packages on monorepo, for example storybook or eslint. 
Especially when building the final image, you don't really want to install those dependencies as all the checks has already passed. 

There is no mention of this command in the migration guide or the rename guide, 
I suppose it was discarded because of its "unreliability" however there is no alternative. Installing hundreds of useless packages is a bit painful.

Currently, as far as my knowledges goes, the only way to achieve that previous behavior is to modify the package.json before running `yarn install`.

Can you recommend a solution and/or add a note about that in the documentation?

Thanks a lot,

**Comment by bodinsamuel** [comment]
thanks for the move @merceyz, not used to this feature 😬

**Comment by arcanis** [comment, marked as answer]
It has been removed. It's not so much because of its instability, but rather because noone used it right, and it wasn't very helpful to begin with. Optional dependencies, as defined by the `optionalDependencies` flag, and that we still support, only have their *build* optional. The packages are still fetched, and still installed - it's just that if they have built steps, then errors that could happen then are downgraded into warnings. As such, not installing those packages at all would be technically incorrect as it would produce a different execution environment in different cases.

Generally speaking we don't really want to support "partial views" of your project in the core unless we can guarantee the result will make sense. So for instance, we have `yarn workspaces focus` because we can somewhat guarantee that even if not all the project deps are installed, you'll at least have all those for the workspace you requested. By contrast, we wouldn't have any way to be sure that a install made via `--ignore-optional` would be correct - so it's place isn't in the core offering.

If you want to reintroduce it, you can do so by either calling `yarn remove` instead of a classic install, or by implementing a plugin similar to `yarn workspaces focus` that would automatically do the `remove` calls in a single operation. I still wouldn't recommend it myself, though.

**Reply by bodinsamuel** [reply]
Make sense.

I think it would deserve a small mention in the documentation. 
And look into the plugin to see if we can implement a similar behaviour 👍🏻 

Thanks for the answer

**Reply by AmirABody** [reply]
Could you find a clean workaround for ignoring optionalDeps during install? any plugins, or anything else?!

**Reply by cellog** [reply]
this argument makes sense, but is insufficient for many cases.

For example, we have a much smaller docker image, for which we do not want to install Cypress because it balloons both installation time by 3 minutes _and_ requires us to store an additional 2GB of stuff that is never used. With `--ignore-optional`, the packages are not installed.

Moving the packages to `devDependencies` simply fails because the Docker image does not have the prerequisites that Cypress requires to install.

Since the recommended approach is to `yarn remove` all the optional deps prior to installation, I guess we will have to do that.

I also noticed that `install --production` is deprecated in favor of `workspaces focus -A --production`. The `--frozen-lockfile` flag (or `--immutable`) is not mentioned, and not supported - can we assume it will fail if the lockfile needs updating?

**Comment by fabioDMFerreira** [comment]
@arcanis any workaround we can use to remove these optional dependencies? I was able to do it manually but it isn't user friendly.

I am having compiling issues because of having optional dependencies in node_modules that have peer dependencies. If I have this specific optional dependency installed I am forced to install a bunch of dependencies I don't need to make the build process work.

**Comment by cmark1302** [comment]
looking for same

**Comment by AmirABody** [comment]
At least, provide a possibility to ignore specific packages during `yarn install`, because I myself used optionalDependencies for this purpose also.
In my own case, I have some pipelines that I don't need super heavy packages like cypress there, so it would be very useful to have such feature implemented in the core.

**Comment by AmirABody** [comment]
Sounds like pretty logical not to ignore optionalDeps during install, since I found that in my own case, when I ran `yarn install --ignore-optional` in my merge request pipeline (to ignore cypress from being installed), my jest tests encountered error, since I was using swc and it has its core binaries as optionalDependencies in the package.
Although ignoring cypress install, declined the step time by 75%, I should consider another solution for the issue.

**Comment by jamie-pate** [comment]
It makes sense to me that we _should_ be able to skip _all_ optional packages. It seems like there are some good use cases to _selectively_ have different execution environments as an optimization which can be e.g. validated by CI running in the environment lacking the optional dependencies....

It would be ok to stamp the documentation with 'Here be dragons/Tread carefully' on this feature imo.

**Comment by ddfridley** [comment]
My issue is that when I push my repo to production, if cypress, storybook and jest are in there, the image is too large, and I get warnings that if it gets much bigger then I won't be able to install it. (or I can pay more for a bigger memory size).   But these packages are not required on the production server. (The build also takes much longer with those packages, which also becomes an issue where my build times out, but I can get around that buy keeping the cache and building more than once.) 

If I could wish for something, we should have buildDependencies and testDependencies.  build dependencies are installed to build the repo/package, but are removed after build, and testDependiencies are  not installed,, except at the top level of the install command, or if "top level" is too tricky, then when install is run with an --install-testDependencies flag.  devDependencies are kind of like buildDependencies except some times they are removed and sometimes they are not and we need something more explicit.
